software/bsg_manycore_lib/bsg_shared_mem.hpp

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+extern "C" {
+#include "bsg_manycore.h"
+#include "bsg_tile_group_barrier.hpp"
+}
+#include <cstdlib>
+#include <cmath>
+
+namespace bsg_manycore {
+    /**
+     * Offset from DMEM[0] = local address - DMEM beginning address (0x1000)
+     *                       =    <12-s>      -   <s-2>    -    00
+     *                             ADDR         Stripe 
+     * s = log2(Stripe Length)
+     *
+     * PREFIX    -    HASH (STRIPE SIZE)    -    UNUSED    -    ADDR    -    Y    -    X    -    Stripe    -    00
+     *  <5>                  <4>            -   <11-x-y>   -  <12 - s>  -   <y>   -   <x>   -    <s-2>     -    <2>
+     * Stripe is lowest bits of offset from local dmem
+     *
+     */
+
+    // log2 is non-constexpr in llvm so we define a custom
+    // constexpr ceil(log2()) function 
+    constexpr uint32_t cilog2(unsigned val) {
+        return val ? 1 + cilog2(val >> 1) : -1;
+    }
+
+    constexpr bool is_powerof2(std::size_t v) {
+        return v && ((v & (v - 1)) == 0);
+    }
```

Comment:
I think these two functions are more general than shared mem

Reply:
Good point! moved them to bsg_manycore.hpp